google-cloud-logging/src/main/java/com/google/cloud/logging/MonitoredResourceUtil.java

```diff
@@ -193,18 +195,47 @@ private static String getAppEngineInstanceName() {
   }
 
   private static List<LoggingEnhancer> getEnhancers(Resource resourceType) {
-    List<LoggingEnhancer> enhancers;
+    List<LoggingEnhancer> enhancers = new ArrayList<>();
     switch (resourceType) {
       // Trace logging enhancer is supported on GAE Flex and Standard.
-      case GaeAppStandard:
       case GaeAppFlex:
-        enhancers = new ArrayList<>();
-        enhancers.add(new TraceLoggingEnhancer());
+        enhancers.add(new LabelLoggingEnhancer(
+            APPENGINE_LABEL_PREFIX, Collections.singletonList(Label.InstanceName)));
+        enhancers.add(new TraceLoggingEnhancer(APPENGINE_LABEL_PREFIX));
+        break;
+      case GaeAppStandard:
+        enhancers.add(new TraceLoggingEnhancer(APPENGINE_LABEL_PREFIX));
         break;
       default:
         enhancers = Collections.emptyList();
         break;
     }
     return enhancers;
   }
+
+  /**
+   *  Adds additional resource-based labels to log entries.
+   *  Labels that can be provided with {@link MonitoredResource.Builder#addLabel(String, String)}
+   *  are restricted to a supported set per resource.
+   *  @see <a href="https://cloud.google.com/logging/docs/api/v2/resource-list">Logging Labels</a>
+   */
+  private static class LabelLoggingEnhancer implements LoggingEnhancer {
+    String prefix;
+    List<Label> labels;
+
+    LabelLoggingEnhancer(String prefix, List<Label> labels) {
```

Comment:
Agree, this was not a good fix given mutability and perf. I've updated this to just build the map in the constructor itself.